Proposal for hwb support in the sass:color module (#2835)
sass/language: proposal/color-4-hwb.md

# HWB Color Functions: Draft 1

*([Issue](https://github.com/sass/sass/issues/2831))*

This proposal adds a new `hwb()` color format to the `sass:color` module, along
with inspection and adjustment options for _whiteness_ and _blackness_.

## Table of Contents

* [Background](#background)
* [Summary](#summary)
  * [Design Decisions](#design-decisions)
* [Semantics](#semantics)
  * [`hwb()`](#hwb)
  * [`whiteness()`](#whiteness)
  * [`blackness()`](#blackness)
  * [`adjust()`](#adjust)
  * [`change()`](#change)
  * [`scale()`](#scale)

## Background

> This section is non-normative.

> The [CSS Color Module level 4][color-4], provides several new CSS formats for

Reviewer: Nit

```suggestion
The [CSS Color Module level 4][color-4] provides several new CSS formats for
```

describing color, but `hwb()` stands out as part of the same `sRGB` color
system that Sass already uses internally.

[color-4]: https://www.w3.org/TR/css-color-4/

## Summary

> This section is non-normative.

> This proposal defines  aSassified versions of the `hwb()` color function

Reviewer: Nit

```suggestion
This proposal defines a Sassified version of the `hwb()` color function
```

- The new functions **will not** accept *special number string* or
  *special variable string* values that can only be resolved in CSS.

Unknown CSS functions are already passed through implicitly. As browser support
grows for the native-CSS functions, authors will be able to migrate on their own
schedule – or use both Sass & CSS functions together to progressively enhance
their designs.

Since all `color.hwb()` output will be processed for rgb output, there is no
need to support shortcuts like `hwb($color, $alpha)`. However, it is reasonable
to support both the space-delimited CSS syntax, as well as the comma-delimited
Sass named-argument syntax.

## Semantics

### `hwb()`

* ```
  hwb($hue, $whiteness, $blackness, $alpha: 1)
  ```

  * If any of `$hue`, `$whiteness`, `$blackness`, or `$alpha` aren't numbers,
    throw an error.

  * If either of `$whiteness` or `$blackness` is not a percentage between 0% and
    100%, throw an error.

  * Let `hue` be `($hue % 360) / 6`.

  * Let `whiteness` and `$blackness` be the respective values of `$whiteness`
    and `$blackness`

  * If the sum of `whiteness` and `blackness` is greater than 100%, normalize

>     both to maintain the same relative ratio, adding up to 100% combined.

Reviewer: Please describe the process of normalizing these in more detail, ideally with a formula.

>   * Let `red`, `green`, and `blue` be the result of converting `hue`,

Reviewer: One thing we haven't explicitly specified yet is what Sass's spec representation is of colors that were originally specified in different formats. In Dart Sass, we store colors as HSL values if they were originally HSL and RGB if they were originally RGB, lazily converting to the other format if that becomes necessary. Presumably we'll do something similar for HWB. This can be visible to the user—we should be explicit about whether, for example, `whiteness(hwb(..., $x, ...))` is guaranteed to return *exactly* `$x` or merely *approximately* `$x`.

I think the actual way you've specified this, treating sRGB values as the internal representation, is reasonable... but we should add an explicit amount of flexibility to enable the sort of optimization Dart Sass does.

You've also got to specify what the function actually returns :smile:.

    `whiteness`, and `blackness` [to RGB][].

  [to RGB]: https://www.w3.org/TR/css-color-4/#hwb-to-rgb

> * ```

Reviewer: If this isn't a global function, why accept the global function color syntax?

  hwb($channels)
  ```

  * If `$channels` is not an unbracketed space-separated list, throw an error.

  * If `$channels` does not includes exactly three elements, throw an error.

  * Let `hue` and `whiteness` be the first two elements of `$channels`

  * If the third element of `$channels` has preserved its status as
    two slash-separated numbers:

    * Let `blackness` be the number before the slash and `alpha` the number
      after the slash.

  * Otherwise:

    * Let `blackness` be the third element of `$channels`.

  * Call `hwb()` with `hue`, `whiteness`, `blackness`, and `alpha` (if it's
    defined) as arguments and return the result.

### `whiteness()`

```
whiteness($color)
```

* If `$color` is not a color, throw an error.

> * Return the `hwb()` whiteness value of `$color`.

Reviewer: I think this warrants more elaboration. Can you link to the algorithm for computing this?

### `blackness()`

```
blackness($color)
```

* If `$color` is not a color, throw an error.
* Return the `hwb()` blackness value of `$color`.

### `adjust()`

> This proposal adds new `$whiteness` and `$blackness` parameters to the `adjust()`

Reviewer: These functions should also be more explicit about what they actually do.

To give some background: because writing a full language spec from scratch would take way too much time, we sort of fill it in piecemeal as it becomes relevant to each new proposal. This means that proposals like this need to do a bit of extra legwork to specify at least enough of the existing behavior to clearly specify the new behavior in relation to it.

forwards-compatible with future CSS changes.

Because the `color.hwb()` function isn't currently intended to directly
implement CSS's native `hwb()` function, it **will not** accept *special number
string* or *special variable string* values that can only be resolved in CSS.
However, for consistency with Sass's `rgb()` and `hsl()` functions it will
support both space-delimited and comma-delimited arguments.

### Future Designs

It's likely that as CSS Color Level 4 matures, `hwb()` will be stabilized and
supported in browsers in one form or another. At this point, Sass will likely
add support for a global `hwb()` function that's compatible with its CSS usage,
including supporting special number and variable strings. The details of this
are left to a future proposal.

## Procedures

### Scaling a Number 

This algorithm takes a number `number`, a value `factor`, and a number `max`.
It's written "scale `<number>` by `<factor>` with a `max` of `<max>`". It
returns a number with a value between 0 and `max` and the same units as
`number`.

* If `factor` isn't a number with unit `%` between `-100%` and `100%`
  (inclusive), throw an error.

> * If `factor > 0%`, return `number + (max - number) * factor / 100%`.

Reviewer 2: What happens if `max < number`?

Reviewer: In practice it never will be, since `number` always comes from a color channel which is guaranteed to be between 0 and `max`. I'll add a note to that effect.

are left to a future proposal.

## Procedures

### Scaling a Number 

This algorithm takes a number `number`, a value `factor`, and a number `max`.
It's written "scale `<number>` by `<factor>` with a `max` of `<max>`". It
returns a number with a value between 0 and `max` and the same units as
`number`.

> Note: in practice, this is only ever called with `number <= max`.

* If `factor` isn't a number with unit `%` between `-100%` and `100%`
  (inclusive), throw an error.

* If `factor > 0%`, return `number + (max - number) * factor / 100%`.

* Otherwise, return `number + number * factor / 100%`.

## Functions

All new functions are part of the `sass:color` built-in module.

### `hwb()`

* ```
  hwb($hue, $whiteness, $blackness, $alpha: 1)
  ```

  * If any of `$hue`, `$whiteness`, `$blackness`, or `$alpha` aren't numbers,
    throw an error.

  * If `$hue` has any units other than `deg`, throw an error.

  * If either of `$whiteness` or `$blackness` don't have unit `%` or aren't
    between `0%` and `100%` (inclusive), throw an error.

  * Let `hue` be `($hue % 360) / 60` without units.

  * Let `whiteness` be `$whiteness / 100%`.

>   * Let `$blackness` be `$blackness / 100%`.

Reviewer 3: ```suggestion
  * Let `blackness` be `$blackness / 100%`.
```

  * If `whiteness + blackness > 1`:

    * Set `whiteness` to `whiteness / (whiteness + blackness)`.

    * Set `blackness` to `blackness / (whiteness + blackness)`.

  * Let `red`, `green`, and `blue` be the result of converting `hue`,

>     `saturation`, and `lightness` [to RGB][].

Reviewer 3: ```suggestion
    `whiteness`, and `blackness` [to RGB][].
```

  * `hwb(hue($color), whiteness($color), blackness($color))` returns a color with
    the same red, green, and blue channels as `$color`.

  * `whiteness($color) + blackness($color) <= 100%`.

  > The specific number returned here is left purposefully open-ended to allow
  > implementations to pursue different strategies for representing color
  > values. For example, one implementation may eagerly convert all colors to
  > RGB channels and convert back when `whiteness()` or `blackness()` is called,
  > where another may keep around their original HWB values and return those
  > as-is.

### `adjust()`

This proposal adds new `$whiteness` and `$blackness` parameters to the `adjust()`
function, and its global `adjust-color()` alias.

```
adjust($color,
  $red: null, $green: null, $blue: null,
  $hue: null, $saturation: null, $lightness: null,
  $whiteness: null, $blackness: null,
  $alpha: null)
```

This function's new definition is as follows:

* If `$color` isn't a color, throw an error.

* Let `alpha` be `$color`'s alpha channel.

* If `$alpha` isn't null:

  * If `$alpha` isn't a number between -1 and 1 (inclusive), throw an error.

>   * Set `alpha` to `alpha + $alpha` clamped between 0 and 255.

Reviewer 3: ```suggestion
  * Set `alpha` to `alpha + $alpha` clamped between 0 and 1.
```

    units otherwise.

  * Let `green` be `$color`'s green channel if `$green` is null or `$green`
    without units otherwise.
  
  * Let `blue` be `$color`'s blue channel if `$blue` is null or `$blue` without
    units otherwise.

  * Return a color with `red`, `green`, `blue`, and `alpha` as the red, green,
    blue, and alpha channels, respectively.

* Otherwise, if either `$saturation` or `$lightness` aren't null:

  * If either `$whiteness` or `$blackness` aren't null, throw an error.

  * If either `$saturation` or `$lightness` aren't either null or numbers
    between 0 and 100 (inclusive), throw an error.

  * Let `hue` be the result of calling `hue($color)` if `$hue` is null, or
    `$hue` otherwise.

  * Let `saturation` be the result of calling `saturation($color)` if
    `$saturation` is null, or `$saturation` otherwise.

  * Let `lightness` be the result of calling `lightness($color)` if
    `$lightness` is null, or `$lightness` otherwise.

  * Return the result of calling [`hsl()`][] with `hue`, `saturation`,
    `lightness`, and `alpha`.

* Otherwise, if either `$hue`, `$whiteness`, or `$blackness` aren't null:

>   * If either `$saturation` or `$lightness` aren't either null or numbers with

Reviewer 3: ```suggestion
  * If either `$whiteness` or `$blackness` aren't either null or numbers with
```

  * If any of `$hue`, `$whiteness`, `$blackness`, or `$alpha` aren't numbers,
    throw an error.

>   * If `$hue` has any units other than `deg`, throw an error.

Author: CSS accepts several different `<angle>` units. I don't know how much they are used (`deg` is certainly the most common), but we might want to match? https://www.w3.org/TR/css3-values/#angle-value

Reviewer: We should support that, but it's tangential enough that I don't think it should be part of this proposal. I've filed https://github.com/sass/sass/issues/2904 to track that.

  * If either of `$whiteness` or `$blackness` don't have unit `%` or aren't
    between `0%` and `100%` (inclusive), throw an error.

  * Let `hue` be `($hue % 360) / 60` without units.

  * Let `whiteness` be `$whiteness / 100%`.

  * Let `$blackness` be `$blackness / 100%`.

  * If `whiteness + blackness > 1`:

    * Set `whiteness` to `whiteness / (whiteness + blackness)`.

    * Set `blackness` to `blackness / (whiteness + blackness)`.

  * Let `red`, `green`, and `blue` be the result of converting `hue`,
    `saturation`, and `lightness` [to RGB][].

  * Set `red`, `green`, and `blue` to their existing values multiplied by 255
    and rounded to the nearest integers.

  * Let `alpha` be the result of [percent-converting][] `$alpha` with a `max` of 1.
  

>   * Return a color with the given `red`, `green`, `blue`, and `alpha` channels.

Author: Returning values from a function is totally overrated 😂

* If `$alpha` isn't null:

  * If `$alpha` isn't a number between -1 and 1 (inclusive), throw an error.

  * Set `alpha` to `alpha + $alpha` clamped between 0 and 255.

* If `$hue` isn't a number or null, throw an error.

* If any of `$red`, `$green`, or `$blue` aren't null:

  * If any of `$hue`, `$saturation`, `$lightness`, `$whiteness`, or `$blackness`
    aren't null, throw an error.

  * If any of `$red`, `$green`, or `$blue` aren't either null or numbers between
    -255 and 255 (inclusive), throw an error.

  * Let `red`, `green`, and `blue` be `$color`'s red, green, and blue channels.

  * If `$red` isn't null, set `red` to `red + $red` clamped between 0 and 255.

  * If `$green` isn't null, set `green` to `green + $green` clamped between 0 and 255.

  * If `$blue` isn't null, set `blue` to `blue + $blue` clamped between 0 and 255.

  * Return a color with `red`, `green`, `blue`, and `alpha` as the red, green,
    blue, and alpha channels, respectively.

* Otherwise, if either `$saturation` or `$lightness` aren't null:

  * If either `$whiteness` or `$blackness` aren't null, throw an error.

  * If either `$saturation` or `$lightness` aren't either null or numbers

>     between -100 and 100 (inclusive), throw an error.

Reviewer 2: Here and below (lines 265, 268, 347): should these have unit `%`?

Reviewer: Unfortunately, this represents the current behavior. We currently ignore units for saturation and lightness no matter what, which isn't ideal but isn't necessarily enough of a problem to warrant a deprecation pass either.
